modules/core/src/main/scala/com.snowplowanalytics/iglu.schemaddl/parquet/Migrations.scala

```diff
@@ -0,0 +1,208 @@
+package com.snowplowanalytics.iglu.schemaddl.parquet
+
+import com.snowplowanalytics.iglu.schemaddl.parquet.Type.{Array, Struct}
+
+/*
+  Parquet schemas are migrated though the `merge.schema = true` loading option in Spark/Databricks.
+  Migrations in this package do not define spark 3.0 style `ATLER TABLE` transformations, similar to Redshift. Instead
+  they are used to categorize schema changes into 2 categories:
+   - Non breaking - Miscellaneous changes that would be handled by the `merge.schema`:
+      * `NullableRequired` - nullable to required change
+      * `NestedKeyAddition` - addition of the new Struct key anywhere except for top level Schema.
+      * `TypeWidening` - Compatible type casting. Such as Int -> Long (cast as BitInt) -> Double.
+      * `TopLevelKeyAddition` - key addition in top level schema
+   - Breaking - Breaking change that would either lead to oen of these outcomes:
+      * `KeyRemoval` historical data loss, such as removing field in Struct. Loading will succeed with data loss.
```

Comment:
I was trying to work out what you mean by historical data loss.  I guess it's this situation...

If a single batch contains BOTH version `1-0-0` and `1-0-1` of a schema in the same batch, then currently the transformer writes out a parquet file based on the `1-0-1` schema.  So `1-0-0` data _in that batch_ will be missing the field.

That's true at the moment.  I'm hoping that's the kind of thing that will get fixed as we improve the transformer to be more aware of schema evolution.

Can we revisit this after making progress on the transformer?

Reply:
If the key is removed, the whole table would evolve a schema to remove the key. Not just a current batch. And it would be forever lost after `OPTIMIZE`.

Reply:
My above comment is wrong.  Missing columns would be merged correctly.